man/man1/xpdtest.1
.\"
.\" $Id: root-config.1,v 1.4 2005/03/21 21:42:21 rdm Exp $
.\"
.TH XPDTEST 1 "Version 5" "ROOT"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
xpdtest \- xproofd test utility
.SH SYNOPSIS
.B xpdtest
.I "[options]"
.SH "DESCRIPTION"
This manual page documents the
.BR xpdtest
program, an utility whose purpose is to test the status of an xproofd daemon running on the local machine.
The target is to run
.B xpdtest
inside a monitor infrastructure, for example
.B monit
(see
.UR http://mmonit.com/monit/
\fIhttp://mmonit.com/monit/\fP
), capable to take some decisions and to do some actions following the result of
.B xpdtest.


.SH OPTIONS
.TP
.B \-\-help, \-h
Gives a short list of options available, and exit
.TP
.B \-t\fP \fR<test>\fP
type of test to be run:
    0       ping the daemon (includes process existence check if pid specified; see below)
    1       ping the daemon and check connection for default user
    2       ping the daemon and check connection for the default user and all recent users

The default user is the one specified in the URL (e.g. 'user' for user@localhost); if no explicit specification is found in the URL, the user under which
.B xpdtest
is run is used; it can be also passed via the environment variable
.B XPDTEST_TEST.
.TP
.B \-u\fP \fR<url>\fP
URL where the xproofd under test responds; it can be also passed via the environment variable
.B XPDTEST_URL
(default is 'localhost:1093').
.TP
.B \-d\fP \fR<sbdir>\fP
sandbox directory used to find out the users of the facility when test is 2; it can be also passed via the environment variable
.B XPDTEST_SBOXDIR
(default is '/tmp/proofbox').
.TP
.B \-s\fP \fR<span>\fP
define the time interval, in minutes, to define 'recent' users when test is 2: only users who connected within this interval are checked; use -1 for infinite;
it can be also passed via the environment variable
.B XPDTEST_TIMESPAN
(default is -1).
.TP
.B \-T\fP \fR<timeout>\fP
define the time, in seconds, waited for the session to start successfully;
it can be also passed via the environment variable
.B XPDTEST_TIMESOUT
(default is 10 seconds).
.TP
.B \-l\fP \fR<logfile>\fP
log file if not screen; deleted if the required test fails unless '-k' is specified (see below);it can be also passed via the environment variable
.B XPDTEST_LOGFILE.
.TP
.B \-p\fP \fR<pidfile>\fP
file with the process ID as a integer number in the first line; if this file is given a quick test on the process existence is done using kill(\fRpid\fP,0);
it can be also passed via the environment variable
.B XPDTEST_PIDFILE.
.TP
.B \-k
keep log file at path given via '-l' in all cases; it can be also passed via the environment variable
.B XPDTEST_KEEP
(default is 'no keep').
.TP
.B \-v
set
.B
gDebug=1
for the underline ROOT calls; it can be also passed via the environment variable
.B XPDTEST_VERBOSE
(default is 'no verbose').
.SH RETURN
.B xpdtest
returns 0 on success and 1 in case of test failure. Errors are printed on the screen or saved into the specified log file.

.SH FILES
The file
.B xpdtest_popen_file
in the temporary directory (typically /tmp) is used to parse the output of
.B TProof::Open().
The file is overwritten by each new attempt; after a run of
.B xpdtest
it contains the result of the last session open attempt.

.SH NOTES
The connection attempt is done in
.B 'masteronly'
mode. For daemons running in
.B 'worker'
mode the attempt will fail with a well defined error message ('Server not allowed to be top master');
since this indicates that the daemon is anyhow responsive, a positive detection of such a pattern in the error message
is interpreted as the attempt having being successful.

.SH "ORIGINAL AUTHORS"
The ROOT team:
.RS
\fBRene Brun\fR and \fBFons Rademakers\fR
.RE
More information can be found at the \fBROOT\fR website:
.UR http://root.cern.ch
\fIhttp://root.cern.ch\fB
.SH "COPYRIGHT"
This library is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation; either version 2.1 of the
License, or (at your option) any later version.
.P
This library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.
.P
You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
.SH AUTHOR
This manual page was written by Gerardo Ganis
<gerardo.ganis@cern.ch>, for the GNU/Linux system (but may be used by
others).
.\"
.\" EOF
.\"


js/changes.md
# JSROOT changelog

## Changes in dev
1. Correctly implement TH2 projections like MERCATOR or PARABOLIC
2. Use https://github.com/georgealways/lil-gui/ instead of dat.GUI
3. Let configure material and scene properties in geom control gui
4. Upgrade three.js r151 -> r155
5. Let toggle vertical/horizontal flag for color palette via context menu
6. Provide "Bring to front" menu command for different objects like pave, box, marker, ...
7. Handle "dark mode" in geom painter - automatically adjust background
8. Change `TF1` and `TF2` drawing - always convert into histogram, support TWebCanvas, handle log scales
9. Add "Superimpose" menu command in hierarchy - let select draw option when append item to pad
10. Support "pol", "cyl", "sph" and "psr" coordinates systems with lego and surf plots
11. Use "col" as default draw option for TH2, "box2" for TH3
12. Support "mollweide" projection for TH2
13. Basic latex support when drawing axes labels and titles in 3D
14. Support orthographic camera for lego and surface plots
15. Implement "tri1", "tri2", "triw" draw options for TGraph2D with Delaunay algorithm
16. Draw axes grids in front of objects - making it equivalent to original ROOT
17. Put `gl` in "devDependencies" of package.json. One can skip it installation with `npm i --production`.
18. Add support of `TPaveClass`
19. Support `inspectN` draw option, allows automatically expand object content to specified level
20. Implement `allfunc` draw option for histograms, force drawing disregard of TF1::kNotDraw bit
21. Use `eslint` for static code checking, add testing of interactive features


## Changes in 7.4.3
1. Fix - correctly use GMT specifier in time format
2. Fix - logical error in `decodeUrl`
3. Fix - member-wise streaming of std::map   #262


## Changes in 7.4.2
1. Fix - unzoom z on lego2 plots
2. Fix - browsing TLists with nullptr inside
3. Fix - check NaN values when performing TTree::Draw()
4. Fix - support standard log function in TF1/TF2


## Changes in 7.4.1
1. Fix - context menu position on lego plots
2. Fix - add missing math functions Chebyshev0 and normalized Gaus
3. Fix - correctly render TPolyLine3D
4. Fix - properly add interactive resize elements for paves and frame
5. Fix - drag and drop handling on tabs layout


## Changes in 7.4.0
1. Upgrade d3.js v7.6.1 -> v7.8.4
2. Upgrade three.js r146 -> r151
3. Support `[cutg]` draw option for TH2
4. Correctly handle `same0` draw option for TH2
5. Fix several issues with axis reverse order, support on lego plots
6. Support more kinds of log scales - ln and logN where N is any positive integer
7. Adjust TAxis title positioning to native ROOT behavior
8. Add interactivity (moving, context menu) for TLine, TBox, TText, TLatex, TDiamond, TGaxis, TASImage
9. Use new gStyle attributes for candle and violin plots
10. Implement autoplace for TLegend, also via context menu
11. Change algorithm of building smooth (bezier) curves
12. Let change physical node visibility in TGeo drawings
13. Use TGaxis attributes from gStyle - fAxisMaxDigits, fStripDecimals and exponent offset
14. Implement "projxy" draw option for TH2 - like projxy3 or projx1_y5
15. Support custom function in TGaxis - when drawn in TWebCanvas
16. Introduce settings.WithCredentials, set xhr.withCredentials = true when submitting HTTP requests
17. Let superimpose TH3 and geo drawings
18. Apply pad draw options like 'gridx' or 'logy' to all subpads
19. Support new TScatter and TAnnotation classes
20. Implement moving and resizing of subpads
21. Implement zooming in the TASImage
22. Let configure position and direction of camera for TGeo, let create URL for that
23. Support labels rotation for simple axis in geometry
24. Support many orthographic cameras with overlayed grid/labels
25. Support InstancedMesh for TGeo drawing, let show really large geometries
26. Implement 'inject=path/script_name.js' url option to inject scripts without emulating of v6
27. Exclude 'HEAD' http request when reading ROOT file, all necessary info can be get from first real HTTP request
28. Provide makeImage function for generation of svg, png and jpeg images in batch and interactively (#257)
29. Implement interactive zoom shifting when middle-mouse button down or single-touch moving
30. Several improvements for touch devices or devices with small displays
31. Remove settings.FrameNDC, use Style.fPadLeft/Right/Top/BottomMargin values instead
32. Fix - rescan sumw2 when update TH1
33. Fix - correct placing for TLegend header
34. Fix - correctly align sub/super scripts in complex TLatex
35. Fix - correctly set visibility level for geo drawing (#258)
36. Fix - use more factor for number of nodes in geo drawing (#258)


## Changes in 7.3.4
1. Fix - failure in normal_cdf calculation
2. Fix - check in TTree::Draw for null buffer
3. Fix - do not rise exception in treeProcess

main/src/xpdtest.cxx
// @(#)root/main:$Id$
// Author: G. Ganis 2012

/*************************************************************************
 * Copyright (C) 1995-2000, Rene Brun and Fons Rademakers.               *
 * All rights reserved.                                                  *
 *                                                                       *
 * For the licensing terms see $ROOTSYS/LICENSE.                         *
 * For the list of contributors see $ROOTSYS/README/CREDITS.             *
 *************************************************************************/

//////////////////////////////////////////////////////////////////////////
//                                                                      //
// xpdtest                                                              //
//                                                                      //
// Program used to test existence and status of a proof daemon          //
//                                                                      //
//////////////////////////////////////////////////////////////////////////
#include <stdio.h>
#include <errno.h>

#include <sys/time.h>
#include <sys/resource.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <signal.h>

#include "Riostream.h"
#include "TSystem.h"
#include "TProof.h"
#include "TString.h"

//______________________________________________________________________________
// Prototypes
int xpd_ping(const char *host, int port);
int getsocket(struct hostent *h, int);
void printhelp();
int recvn(int sock, void *buffer, int length);
int sendn(int sock, const void *buffer, int length);
int proof_open(const char *master, long to = 10);
int parse_args(int argc, char **argv,
               TString &url, TString &sboxdir, time_t &span, int &test,
               TString &logfile, bool &keep, bool &verbose, long &to,
               TString &pidfile);
void printhelp();
int set_timer(bool on = 1, long to = 10);

//______________________________________________________________________________
// The client request structure
typedef struct {
   int first;
   int second;
   int third;
   int fourth;
   int fifth;
} clnt_HS_t;

//______________________________________________________________________________
// The body received after the first handshake's header
typedef struct {
   int msglen;
   int protover;
   int msgval;
} srv_HS_t;

////////////////////////////////////////////////////////////////////////////////
///  PROOF daemon test program.
///  Syntax
///          xpdtest  `<url>` `<sandbox_dir>` `<time_span>`
///
///          `<url>`            URL to test; default 'localhost:1093'
///          `<sandbox_dir>`    directory with users sandboxes; used to find out
///                           users to check connection; default '/tmp/proofbox'
///          `<time_span>`      check only users whose latest activity was within
///                           'time_span' minutes; use -1 to check all users;
///                           default -1.
///
///  Exits 0 on success, 1 on error

int main(int argc, char **argv)
{
   TString url, sboxdir, logfile, pidfile;
   time_t span = -1;
   int test = 0;
   bool keep = 0, verbose = 0;
   long timeout = -1;
   int rc = parse_args(argc, argv, url, sboxdir, span, test, logfile, keep,
                       verbose, timeout, pidfile);
   if (rc < 0) {
      fprintf(stderr, "xpdtest: parse_args failure\n");
      gSystem->Exit(1);
   } else if (rc > 0) {
      gSystem->Exit(0);
   }
   gDebug = (verbose) ? 1 : 0;

   rc = 0;

   // Set up log file if required
   RedirectHandle_t redirH;
   if (!logfile.IsNull()) {
      gSystem->RedirectOutput(logfile, "w", &redirH);
   }

   // Extract process ID, if a file has been passed
   if (!pidfile.IsNull()) {
      std::fstream infile(pidfile.Data(), std::ios::in);
      if (infile.is_open()) {
         TString line;
         line.ReadToDelim(infile);
         line.Remove(TString::kTrailing, '\n');
         if (line.IsDigit()) {
            pid_t pid = (pid_t) line.Atoi();
            if (kill(pid, 0) != 0) {
               fprintf(stderr, "xpdtest: process '%d' does not exist\n", (int) pid);
               rc = 1;
            }
         } else {
            fprintf(stderr, "xpdtest: pId file does not contain pid in expected form (first line: %s)\n", line.Data());
            rc = 1;
         }
      } else {
         fprintf(stderr, "xpdtest: pId file '%s' could not be opened\n", pidfile.Data());
         rc = 1;
      }
   }

   // Check sandbox dir
   FileStat_t fst;
   if (rc == 0 && test > 1) {
      if (gSystem->GetPathInfo(sboxdir, fst) != 0) {
         fprintf(stderr, "xpdtest: stat failure for '%s'\n", sboxdir.Data());
         rc = 1;
      }
      if (rc == 0 && !R_ISDIR(fst.fMode)) {
         fprintf(stderr, "xpdtest: '%s' is not a directory\n", sboxdir.Data());
         rc = 1;
      }
   }

   // Setup URL
   TUrl u;
   TString defusr;
   if (rc == 0) {
      u.SetUrl(url.Data());
      defusr = u.GetUser();
      if (defusr.IsNull()) {
         UserGroup_t *pw = gSystem->GetUserInfo();
         if (pw) {
            defusr = pw->fUser;
            delete pw;
         }
      }
      if (!url.BeginsWith(u.GetProtocol())) {
         if (u.GetPort() == 80 && !strcmp(u.GetProtocol(), "http")) u.SetPort(1093);
         u.SetProtocol("proof");
      }
   }

   // Do ping
   if (rc == 0) {
      set_timer(1, timeout);
      if ((rc = xpd_ping(u.GetHost(), u.GetPort())) != 0)
         fprintf(stderr, "xpdtest: failure pinging '%s'\n", url.Data());
      set_timer(0);
   }

   if (test > 0) {
      // Do TProof::Open in "masteronly" mode
      if (rc == 0) {
         if (!logfile.IsNull()) gSystem->RedirectOutput(0, 0, &redirH);
         rc = proof_open(u.GetUrl(), timeout);
         if (!logfile.IsNull()) gSystem->RedirectOutput(logfile, "a", &redirH);
         if (rc != 0)
            fprintf(stderr, "xpdtest: TProof::Open failure for default user '%s'\n", u.GetUrl());
      }

      if (test > 1) {
         if (rc == 0) {

            // Scan sandbox dir
            time_t now = time(0);
            void *dirp = gSystem->OpenDirectory(sboxdir.Data());
            const char *ent = 0;
            TString dent;
            while ((ent = gSystem->GetDirEntry(dirp))) {
               if (!strcmp(ent, "..") || !strcmp(ent, ".")) continue;
               if (defusr == ent) continue;
               FileStat_t st;
               if (span > 0) {
                  dent.Form("%s/%s", sboxdir.Data(), ent);
                  if (gSystem->GetPathInfo(dent, st) != 0) {
                     fprintf(stderr, "xpdtest: stat failure for '%s'\n", dent.Data());
                     rc = 1;
                     break;
                  }
               }
               if (span < 0 || st.fMtime > (now - span)) {
                  u.SetUser(ent);
                  fprintf(stderr, "proof_open: url: '%s'\n", u.GetUrl());
                  if (!logfile.IsNull()) gSystem->RedirectOutput(0, 0, &redirH);
                  rc = proof_open(u.GetUrl(), timeout);
                  if (!logfile.IsNull()) gSystem->RedirectOutput(logfile, "a", &redirH);
                  if (rc != 0) {
                     fprintf(stderr, "xpdtest: failure scanning sandbox dir '%s'\n", sboxdir.Data());
                     break;
                  }
               }
            }
            gSystem->FreeDirectory(dirp);
         }
      }
   }
   if (!logfile.IsNull()) {
      gSystem->RedirectOutput(0, 0, &redirH);
      if (rc == 0 && !keep) gSystem->Unlink(logfile);
   }
   gSystem->Exit(rc);
}

// Auxilliary functions

////////////////////////////////////////////////////////////////////////////////
/// Extract control info from arguments

int parse_args(int argc, char **argv,
               TString &url, TString &sboxdir, time_t &span, int &test,
               TString &logfile, bool &keep, bool &verbose, long &to, TString &pidfile)
{
   url = "localhost:1093";
   sboxdir = "/tmp/proofbox";
   span = -1;
   test = 0;
   logfile= "";
   keep = 0;

   // Check environment settings first
   if (getenv("XPDTEST_URL")) {
      url = getenv("XPDTEST_URL");
   }
   if (getenv("XPDTEST_SBOXDIR")) {
      sboxdir = getenv("XPDTEST_SBOXDIR");
   }
   if (getenv("XPDTEST_TIMESPAN")) {
      errno = 0;
      long xspan = strtol(getenv("XPDTEST_TIMESPAN"), 0, 10);
      if (errno == 0) span = (time_t) xspan;
   }
   if (getenv("XPDTEST_TEST")) {
      errno = 0;
      long xtest = strtol(getenv("XPDTEST_TEST"), 0, 10);
      if (errno == 0 && xtest >= 0 && xtest <= 2) sscanf(getenv("XPDTEST_TEST"), "%d", &test);
   }
   if (getenv("XPDTEST_LOGFILE")) {
      logfile = getenv("XPDTEST_LOGFILE");
   }
   if (getenv("XPDTEST_KEEP")) {
      keep = 1;
   }
   if (getenv("XPDTEST_VERBOSE")) {
      verbose = 1;
   }
   if (getenv("XPDTEST_TIMEOUT")) {
      errno = 0;
      long xto = strtol(getenv("XPDTEST_TIMEOUT"), 0, 10);
      if (errno == 0 && xto > 0) sscanf(getenv("XPDTEST_TIMEOUT"), "%ld", &to);
   }
   if (getenv("XPDTEST_PIDFILE")) {
      pidfile = getenv("XPDTEST_PIDFILE");
   }

   // -u url -d sboxdir -s span -t test
   for (int i = 1; i < argc; i++) {
      if (argv[i]) {
         if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help")) {
            printhelp();
            return 1;
         } else if (!strcmp(argv[i], "-u")) {
            if (argv[++i]) {
               url = argv[i];
            } else {
               fprintf(stderr, "parse_args: '-u' requires the specification of the URL!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-d")) {
            if (argv[++i]) {
               sboxdir = argv[i];
            } else {
               fprintf(stderr, "parse_args: '-d' requires the specification of the sandbox directory!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-l")) {
            if (argv[++i]) {
               logfile = argv[i];
            } else {
               fprintf(stderr, "parse_args: '-l' requires the specification of the logfile!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-k")) {
            keep = 1;
         } else if (!strcmp(argv[i], "-s")) {
            if (argv[++i]) {
               errno = 0;
               long xspan = strtol(argv[i], 0, 10);
               if (errno == 0 && xspan > 0) {
                  span = (time_t) xspan;
               } else {
                  fprintf(stderr, "parse_args: time span must be a positive integer! (argv: %s, xspan: %ld)\n",
                                  argv[i], xspan);
                  printhelp();
                  return -1;
               }
            } else {
               fprintf(stderr, "parse_args: '-s' requires the specification of the time span!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-t")) {
            if (argv[++i]) {
               errno = 0;
               long xtest = strtol(argv[i], 0, 10);
               if (errno == 0 && xtest >= 0 && xtest <= 2) {
                  sscanf(argv[i], "%d", &test);
               } else {
                  fprintf(stderr, "parse_args: test must be an integer in [0,2]! (argv: %s, xtest: %ld)\n",
                                  argv[i], xtest);
                  printhelp();
                  return -1;
               }
            } else {
               fprintf(stderr, "parse_args: '-t' requires the specification of the test!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-T")) {
            if (argv[++i]) {
               errno = 0;
               long xto = strtol(argv[i], 0, 10);
               if (errno == 0 && xto > 0) {
                  to = (time_t) xto;
               } else {
                  fprintf(stderr, "parse_args: timeout must be a positive integer! (argv: %s, xto: %ld)\n",
                                  argv[i], xto);
                  printhelp();
                  return -1;
               }
            } else {
               fprintf(stderr, "parse_args: '-T' requires the specification of the timeout (in secs)!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-p")) {
            if (argv[++i]) {
               pidfile = argv[i];
            } else {
               fprintf(stderr, "parse_args: '-p' requires the specification of the file with the process ID!\n");
               printhelp();
               return -1;
            }
         } else if (!strcmp(argv[i], "-v")) {
            verbose = 1;
         }
      }
   }
   // Done
   return 0;
}

////////////////////////////////////////////////////////////////////////////////
/// Help function

void printhelp()
{
   fprintf(stderr, "\n");
   fprintf(stderr, "   xpdtest: test xproofd service on (remote) host\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "   Syntax:\n");
   fprintf(stderr, "            xpdtest [-u url] [-t test] [-d sbdir] [-s span] [-T timeout] [-l logfile]\n");
   fprintf(stderr, "                    [-p pidfile] [-k] [-v] [-h|--help]\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "   url:     URL where the xproofd under test responds [localhost:1093]\n");
   fprintf(stderr, "   test:    type of test [0]\n");
   fprintf(stderr, "                 0   ping\n");
   fprintf(stderr, "                 1   '0' + check connection for default user\n");
   fprintf(stderr, "                 2   '1' + check connection for all recent users\n");
   fprintf(stderr, "   sbdir:   sandbox directory used to find out the users of the facility\n");
   fprintf(stderr, "            when test is 2 ['/tmp/proofbox']\n");
   fprintf(stderr, "   span:    define the time interval to define 'recent' users when test\n");
   fprintf(stderr, "            is 2: only users who connected within this interval are checked;\n");
   fprintf(stderr, "            use -1 for infinite [-1]\n");
   fprintf(stderr, "   timeout: max time waited for a successful session start in seconds [10 secs]\n");
   fprintf(stderr, "   logfile: log file if not screen; deleted if the test fails unless '-k' is\n");
   fprintf(stderr, "            specified (see below) [terminal]\n");
   fprintf(stderr, "   pidfile: file with the daemon process id; if specified, an initial test of the\n");
   fprintf(stderr, "            process existence is done (using kill(pid, 0))\n");
   fprintf(stderr, "   -k:      keep log file at path given via '-l' in all cases [do not keep]\n");
   fprintf(stderr, "   -v:      set gDebug = 1 for ROOT\n");
   fprintf(stderr, "   -h, --help: print this screen\n");
   fprintf(stderr, "\n");
   fprintf(stderr, "   Return:  0   the required test succeeded\n");
   fprintf(stderr, "            1   the test failed\n");
   fprintf(stderr, "\n");
}

////////////////////////////////////////////////////////////////////////////////
/// Test proof open.
/// Return 0 on success, 1 on failure

int proof_open(const char *master, long to)
{
   int rc = 0;
   RedirectHandle_t rh;
   TString popenfile = TString::Format("%s/xpdtest_popen_file", gSystem->TempDirectory());
   gSystem->RedirectOutput(popenfile, "w", &rh);
   set_timer(1, to);
   TProof *p = TProof::Open(master, "masteronly");
   set_timer(0);
   gSystem->RedirectOutput(0, 0, &rh);
   if (!p || (p && !p->IsValid())) {
      TMacro m;
      m.ReadFile(popenfile);
      if (!m.GetLineWith("Server not allowed to be top master")) rc = 1;
   }
   if (p) delete p;
   return rc;
}

////////////////////////////////////////////////////////////////////////////////
/// If we are called it means that we failed

static void handle_sigalarm(int)
{
   gSystem->Exit(1);
}

////////////////////////////////////////////////////////////////////////////////
/// Do nothing

static void ignore_signal(int)
{
}

////////////////////////////////////////////////////////////////////////////////
/// Start an asynchronous firing after 'to' seconds
/// Return
///        0 on success
///        1 otherwise

int set_timer(bool on, long to)
{
   struct itimerval itv;
   bool sett = (on && to > 0) ? 1 : 0;
   itv.it_value.tv_sec     = (sett) ? time_t(to) : time_t(0);
   itv.it_value.tv_usec    = 0;
   itv.it_interval.tv_sec  = 0;
   itv.it_interval.tv_usec = 0;
   // Enable/disable handling of the related signal
   if (sett) {
      signal(SIGALRM, handle_sigalarm);
   } else {
      signal(SIGALRM, ignore_signal);
   }
   errno = 0;
   return setitimer(ITIMER_REAL, &itv, 0);
}

////////////////////////////////////////////////////////////////////////////////
/// Non-blocking check for a PROOF service at 'url'
/// Return
///        0 if a XProofd daemon is listening at 'url'
///        1 otherwise

int xpd_ping(const char *host, int port)
{
   // Check arguments
   if (!host || (host && strlen(host) <= 0)) {
      fprintf(stderr,"xpd_ping: host must be given!\n");
      return 1;
   }

   struct hostent *h = gethostbyname(host);
   if (!h) {
      fprintf(stderr,"xpd_ping: unknown host '%s'\n", host);
      return 1;
   }

   // Get socket and listen to it
   int sd = getsocket(h,port);
   if (sd == -1) {
      fprintf(stderr,"xpd_ping: problems creating socket ... exit\n");
      return 1;
   }

   // Send the first bytes
   clnt_HS_t initHS;
   memset(&initHS, 0, sizeof(initHS));
   int len = sizeof(initHS);
   initHS.third  = (int)htonl((int)1);
   if (sendn(sd, &initHS, len) != len) {
      fprintf(stderr,"xpd_ping: problems sending first set of handshake bytes\n");
      close(sd);
      return 1;
   }

   // These 8 bytes are need by 'rootd/proofd' and discarded by XRD/XPD
   int dum[2];
   dum[0] = (int)htonl((int)4);
   dum[1] = (int)htonl((int)2012);
   if (sendn(sd, &dum[0], sizeof(dum)) != sizeof(dum)) {
      fprintf(stderr,"xpd_ping: problems sending second set of handshake bytes\n");
      close(sd);
      return 1;
   }

   // Read first server response
   int type;
   len = sizeof(type);
   int nr = 0;
   if ((nr = recvn(sd, &type, len)) != len) { // 4 bytes
      fprintf(stderr, "xpd_ping: 1st: wrong number of bytes read: %d (expected: %d)\n",
                      nr, len);
      close(sd);
      return 1;
   }

   // To host byte order
   type = ntohl(type);
   // Check if the server is the eXtended proofd
   if (type == 0) {
      srv_HS_t xbody;
      len = sizeof(xbody);
      if ((nr = recvn(sd, &xbody, len)) != len) { // 12(4+4+4) bytes
         fprintf(stderr, "xpd_ping: 2nd: wrong number of bytes read: %d (expected: %d)\n",
                         nr, len);
         close(sd);
         return 1;
      }
      xbody.protover = ntohl(xbody.protover);
      xbody.msgval = ntohl(xbody.msglen);
      xbody.msglen = ntohl(xbody.msgval);

   } else if (type == 8) {
      // Standard proofd
      fprintf(stderr, "xpd_ping: server is PROOFD\n");
      close(sd);
      return 1;
   } else {
      // We don't know the server type
      fprintf(stderr, "xpd_ping: unknown server type: %d\n", type);
      close(sd);
      return 1;
   }
   // Cleanup
   close(sd);
   // Done
   return 0;
}

////////////////////////////////////////////////////////////////////////////////

int getsocket(struct hostent *h, int port)
{
   int sd, rc;
   struct sockaddr_in localAddr = sockaddr_in();
   struct sockaddr_in servAddr = sockaddr_in();

   servAddr.sin_family = h->h_addrtype;
   memcpy((char *) &servAddr.sin_addr.s_addr, h->h_addr_list[0], h->h_length);
   servAddr.sin_port = htons(port);

   /* create socket */
   sd = socket(AF_INET, SOCK_STREAM, 0);
   if(sd < 0) {
      perror("cannot open socket ");
      return -1;
   }

   /* bind any port number */
   localAddr.sin_family = AF_INET;
   localAddr.sin_addr.s_addr = htonl(INADDR_ANY);
   localAddr.sin_port = htons(0);

   rc = bind(sd, (struct sockaddr *) &localAddr, sizeof(localAddr));
   if(rc < 0) {
      perror("error ");
      close(sd);
      return -1;
   }

   /* connect to server */
   rc = connect(sd, (struct sockaddr *) &servAddr, sizeof(servAddr));
   if(rc < 0) {
      perror("cannot connect ");
      close(sd);
      return -1;
   }

   return sd;

}

////////////////////////////////////////////////////////////////////////////////
/// Send exactly length bytes from buffer.

int sendn(int sock, const void *buffer, int length)
{
   if (sock < 0) return -1;

   int n, nsent = 0;
   const char *buf = (const char *)buffer;

   for (n = 0; n < length; n += nsent) {
      if ((nsent = send(sock, buf+n, length-n, 0)) <= 0) {
         perror("problems sending ");
         return nsent;
      }
   }

   return n;
}

////////////////////////////////////////////////////////////////////////////////
/// Receive exactly length bytes into buffer. Returns number of bytes
/// received. Returns -1 in case of error.

int recvn(int sock, void *buffer, int length)
{
   if (sock < 0) return -1;

   int n, nrecv = 0;
   char *buf = (char *)buffer;

   for (n = 0; n < length; n += nrecv) {
      while ((nrecv = recv(sock, buf+n, length-n, 0)) == -1 && errno == EINTR)
         errno = 0;   // probably a SIGCLD that was caught
      if (nrecv < 0) {
         perror("problems receiving ");
         return nrecv;
      } else if (nrecv == 0)
         break;         // EOF
   }

   return n;
}
